### ISSUE TITLE
"Could not find platform independent libraries"

### ISSUE THREAD

**Issue by cmacrae** [radekg/terraform-provisioner-ansible#48]
Hey @radekg 👋   

Thank you for this awesome plugin!  
Unfortunately, I'm running into an issue with it.  

### Environment/configuration
Client (terraform & ansible): macOS 10.12.6  
Remote (provisioning target): OpenBSD 6.1  
Terraform version: `v0.11.6`  
Ansible version: `2.4.2.0`

### Issue info
I'm trying to provision an OpenBSD system using a playbook that works absolutely fine if I execute it from the shell against the exact machine that's been provisioned by Terraform.  
However, if it's done as a `provision` step, using your plugin, these are the results I get:  
```
triton_machine.vpn: Provisioning with 'ansible'...
triton_machine.vpn (ansible): Generating temporary ansible inventory...
triton_machine.vpn (ansible): Writing temporary ansible inventory to '/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/temporary-ansible-inventory182202682'...
triton_machine.vpn (ansible): Ansible inventory written.
triton_machine.vpn (ansible): Executing: ["/bin/sh" "-c" "ssh-keyscan -p 22 172.19.2.26 2>/dev/null | head -n1 > \"/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/7467afe4-f8d1-4222-b7ff-c1fb7fb90c03\""]
triton_machine.vpn (ansible): running local command: ANSIBLE_FORCE_COLOR=true ansible-playbook ../../ansible/plays/ipsec.yml --inventory-file='/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/temporary-ansible-inventory182202682' --extra-vars='{"ansible_bash_interpreter":"/usr/local
/bin/bash","ansible_python_interpreter":"/usr/local/bin/python2.7"}' --forks=5 --vault-password-file='../../ansible/pass.sh' --user='root' --ssh-extra-args='-p 22 -o UserKnownHostsFile=/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/7467afe4-f8d1-4222-b7ff-c1fb7fb90c03 -o ConnectTimeout=10 -o ConnectionAttempts=10'
triton_machine.vpn (ansible): Executing: ["/bin/sh" "-c" "ANSIBLE_FORCE_COLOR=true ansible-playbook ../../ansible/plays/ipsec.yml --inventory-file='/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/temporary-ansible-inventory182202682' --extra-vars='{\"ansible_bash_interpreter\":\"/u
sr/local/bin/bash\",\"ansible_python_interpreter\":\"/usr/local/bin/python2.7\"}' --forks=5 --vault-password-file='../../ansible/pass.sh' --user='root' --ssh-extra-args='-p 22 -o UserKnownHostsFile=/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/7467afe4-f8d1-4222-b7ff-c1fb7fb90c03
 -o ConnectTimeout=10 -o ConnectionAttempts=10'"]

triton_machine.vpn (ansible): PLAY [vpn_servers] *************************************************************

triton_machine.vpn (ansible): TASK [Ensure sysctl params are set] ********************************************
triton_machine.vpn (ansible): failed: [172.19.2.26] (item=ip.forwarding) => {"changed": false, "item": "ip.forwarding", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTH
ONHOME to <prefix>[:<exec_prefix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
triton_machine.vpn (ansible): failed: [172.19.2.26] (item=esp.enable) => {"changed": false, "item": "esp.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME
 to <prefix>[:<exec_prefix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
triton_machine.vpn (ansible): failed: [172.19.2.26] (item=ah.enable) => {"changed": false, "item": "ah.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME t
o <prefix>[:<exec_prefix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
triton_machine.vpn (ansible): failed: [172.19.2.26] (item=ipcomp.enable) => {"changed": false, "item": "ipcomp.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTH
ONHOME to <prefix>[:<exec_prefix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
triton_machine.vpn (ansible):   to retry, use: --limit @/Users/cmacrae/code/git/ops/ansible/plays/ipsec.retry

triton_machine.vpn (ansible): PLAY RECAP *********************************************************************
triton_machine.vpn (ansible): 172.19.2.26                : ok=0    changed=0    unreachable=0    failed=1


Error: Error applying plan:

1 error(s) occurred:

* triton_machine.vpn: Error running command 'ANSIBLE_FORCE_COLOR=true ansible-playbook ../../ansible/plays/ipsec.yml --inventory-file='/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/temporary-ansible-inventory182202682' --extra-vars='{"ansible_bash_interpreter":"/usr/local/bin/bas
h","ansible_python_interpreter":"/usr/local/bin/python2.7"}' --forks=5 --vault-password-file='../../ansible/pass.sh' --user='root' --ssh-extra-args='-p 22 -o UserKnownHostsFile=/var/folders/sn/f0p54j5x3l926vy3br_h1c3r0000gn/T/7467afe4-f8d1-4222-b7ff-c1fb7fb90c03 -o ConnectTimeout=10
 -o ConnectionAttempts=10'': exit status 2. Output:
PLAY [vpn_servers] *************************************************************

TASK [Ensure sysctl params are set] ********************************************
failed: [172.19.2.26] (item=ip.forwarding) => {"changed": false, "item": "ip.forwarding", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME to <prefix>[:<exec_pref
ix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
failed: [172.19.2.26] (item=esp.enable) => {"changed": false, "item": "esp.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME to <prefix>[:<exec_prefix>]\r
\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
failed: [172.19.2.26] (item=ah.enable) => {"changed": false, "item": "ah.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME to <prefix>[:<exec_prefix>]\r\n
ImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
failed: [172.19.2.26] (item=ipcomp.enable) => {"changed": false, "item": "ipcomp.enable", "module_stderr": "Shared connection to 172.19.2.26 closed.\r\n", "module_stdout": "Could not find platform independent libraries <prefix>\r\nConsider setting $PYTHONHOME to <prefix>[:<exec_pref
ix>]\r\nImportError: No module named site\r\n", "msg": "MODULE FAILURE", "rc": 0}
        to retry, use: --limit @/Users/cmacrae/code/git/ops/ansible/plays/ipsec.retry

PLAY RECAP *********************************************************************
172.19.2.26                : ok=0    changed=0    unreachable=0    failed=1
```

### `provisioner` expression being used
```
  provisioner "ansible" {
    local = "yes"
    vault_password_file = "../../ansible/pass.sh"
    hosts = ["${self.primaryip}"]
    groups = ["vpn_servers"]
    plays {
      playbook = "../../ansible/plays/ipsec.yml"
      extra_vars {
	ansible_python_interpreter = "/usr/local/bin/python2.7"
	ansible_bash_interpreter = "/usr/local/bin/bash"
      }
    }
  }
```

### Conjecture
It seems to be a problem with the local execution of Ansible.  
If I run the **exact** command produced by your provisioner in the shell, it all works fine.  
It's only when the execution is being called by the plugin that it's failing as above.  
The only thing I can think of that might be interfering is the means by which I have Ansible installed o my client machine: I use Nix for system management on my macOS machines.  
The "binaries" provided by the Ansible package are wrapped (as many other packages are) like so:
```
cmacrae $ cat $(which ansible-playbook)                                                                                            ~
#! /nix/store/h6815iv0gcn7hq1nk6qc1rzsv7x9gcz7-bash-4.4-p12/bin/bash -e
export PATH='/nix/store/573lgkld523h7hh5djfzfl8p66fcxawc-python-2.7.14/bin:/nix/store/9zr862kirdabm620qcni5rgi4glyfa0y-python2.7-ansible-2.4.2.0/bin:/nix/store/smgs43b3mnrja89f0g5dcvl0haf8pqyg-python2.7-setuptools-38.4.0/bin:/nix/store/yp8y4hfdbxf2dbfvgs8k0qcx67jzjsay-python2.7-boto-2.47.0/bin:/nix/store/nn6kp8jknb0al4yhxa39vp9bhv4ssbvp-python2.7-chardet-3.0.4/bin:/nix/store/lma2f5yk6aak4z6jsm7wvvm6idpx07fm-python2.7-netaddr-0.7.19/bin'${PATH:+':'}$PATH
export PYTHONNOUSERSITE='true'
exec -a "$0" "/nix/store/9zr862kirdabm620qcni5rgi4glyfa0y-python2.7-ansible-2.4.2.0/bin/.ansible-playbook-wrapped"  "${extraFlagsArray[@]}" "$@"
```

I think that maybe the `PYTHONNOUSERSITE` is important here.  
So, I was thinking that maybe somehow, when `ansible-playbook` is being called by the plugin, it's foregoing the wrapper script above, and `PYTHONNOUSERSITE` isn't getting set?  

Thank you for your time. Let me know if there's any more information I can provide to assist with debugging. Any help on this is greatly appreciated!

**Comment by radekg** [comment]
Hey @cmacrae, thanks for the report. What happens if you `export PYTHONNOUSERSITE='true'` before `terraform apply`. Eventually `PYTHONNOUSERSITE='true' terraform apply`?

**Comment by cmacrae** [comment]
Sorry I should've mentioned: I tried these already, but unfortunately they didn't help.

**Comment by radekg** [comment]
Would you be so kind and try the following branch? https://github.com/radekg/terraform-provisioner-ansible/tree/PYTHONNOUSERSITE

**Comment by cmacrae** [comment]
Thanks for pushing that up - I just tried it, and unfortunately it's the same problem.  
That said, now that I know where the relevant code is, I can play around and see if there's a solution I can come up with.  

One thing I did notice:  
```
triton_machine.vpn (ansible): Executing: ["/bin/sh" "-c" "ANSIBLE_FORCE_COLOR=true PYTHONNOUSERSITE=true ansible-playbook...
```

I can see it's executing in a bash subshell (`bash -c`), I wonder if that has something to do with it.  
Is this something you implemented? I'll have a dig through the code

**Comment by radekg** [comment]
`/bin/sh` is executed by terraform `local-exec` provisioner: https://github.com/radekg/terraform-provisioner-ansible/blob/PYTHONNOUSERSITE/provisioner.go#L1046. This should not be a problem.

On the second look at the error message: `No module named site`, maybe `ansible-playbook` is not picking up your modules when running via terraform?

**Comment by cmacrae** [comment]
> On the second look at the error message: `No module named site`, maybe `ansible-playbook` is not picking up your modules when running via terraform?  

Yep, that certainly seems to be the case. Just trying to work out why, as it works fine from the shell

**Comment by radekg** [comment]
It definitely looks like no env variables are forwarded to `ansible-playbook`. Do you have any `ANSIBLE_...` env vars set by default in your environment?

**Comment by cmacrae** [comment]
I think you're right, `ansible-playbook` when called by Terraform doesn't seem to be inheriting any of the environment variables. Do you know if locally executed commands should be inheriting the environment variables from the shell they're called in? Or would that be considered unclean/troublesome?  

No, I don't have any `ANSIBLE_...` vars set. In fact looking at `env`, I don't see any that I would suspect to be interfering with Ansible or any Python processes for that matter

**Comment by radekg** [comment]
What is the content of `/nix/store/9zr862kirdabm620qcni5rgi4glyfa0y-python2.7-ansible-2.4.2.0/bin/.ansible-playbook-wrapped` ?

**Comment by cmacrae** [comment]
It's `ansible-playbook` (as on any other system) with the following modifications:  

The shebang is:  
`#!/nix/store/573lgkld523h7hh5djfzfl8p66fcxawc-python-2.7.14/bin/python2.7`  

These site directories added explicitly:  
```
import sys;import site;import functools;sys.argv[0] = '/nix/store/9zr862kirdabm620qcni5rgi4glyfa0y-python2.7-ansible-2.4.2.0/bin/ansible-playbook';functools.reduce(lambda k, p: site.addsitedir(p, k), ['/nix/store/9zr862kirdabm620qcni5rgi4glyfa0y-python2.7-ansible-2.4.2.0/lib/python2.7/site-packages','/nix/store/v41c315d8cpcnrfgqbl8z60bh2wkxzqx-python2.7-pycrypto-3.4.9/lib/python2.7/site-packages','/nix/store/kxz6a7xnckf3n40x9fbzl7z3sxc8y021-python2.7-pycryptodome-3.4.9/lib/python2.7/site-packages','/nix/store/smgs43b3mnrja89f0g5dcvl0haf8pqyg-python2.7-setuptools-38.4.0/lib/python2.7/site-packages','/nix/store/kh5kbcc6lqmb5sqibji6j7fq934y8n9q-python2.7-paramiko-2.1.1/lib/python2.7/site-packages','/nix/store/almmd5b7mmlpnqaypp8d3abh4w3lrki8-python2.7-idna-2.6/lib/python2.7/site-packages','/nix/store/z2smf1p77al5673s97avd0dwswpg2n7h-python2.7-asn1crypto-0.24.0/lib/python2.7/site-packages','/nix/store/2wz70r4ml703fhg7jvgnc63lv8hswxs8-python2.7-packaging-16.8/lib/python2.7/site-packages','/nix/store/4n2fm5nn6nwx2gxghbd3p9kqynk285sb-python2.7-pyparsing-2.2.0/lib/python2.7/site-packages','/nix/store/1lw6llvmfylxicpgma074fd38n5wjafg-python2.7-six-1.11.0/lib/python2.7/site-packages','/nix/store/dx49x3b52cpfakc5k0pgcisiffxdnr4y-python2.7-enum34-1.1.6/lib/python2.7/site-packages','/nix/store/mm74swzv5n9ac3gaf6cpl812nv8ihabi-python2.7-ipaddress-1.0.18/lib/python2.7/site-packages','/nix/store/r8rjp67js1jq98njv5q3v9qpy96sa9sa-python2.7-pycparser-2.14/lib/python2.7/site-packages','/nix/store/pv19bcg872lqrm0pa8w0kcny6dq8irzw-python2.7-cffi-1.11.4/lib/python2.7/site-packages','/nix/store/rq4003s77w8mj748ilczxv358dmh47wp-python2.7-cryptography-2.1.4/lib/python2.7/site-packages','/nix/store/y78c7phqgy9i3s9xxdvhqrdvz83h8kcx-python2.7-pyasn1-0.4.2/lib/python2.7/site-packages','/nix/store/i7gdzbac7zbkqmpimb55zbjnvb85bgp4-python2.7-Jinja2-2.10/lib/python2.7/site-packages','/nix/store/y2arx8jqlns33s2pzsqnqagyqfcz5b4f-python2.7-markupsafe-1.0/lib/python2.7/site-packages','/nix/store/6bdxgh01ff6mmj6nci50n6ax4aw01z51-python2.7-PyYAML-3.12/lib/python2.7/site-packages','/nix/store/7qzsrbg49gid4sb86xfj8hvl3xk918z5-python2.7-httplib2-0.9.2/lib/python2.7/site-packages','/nix/store/yp8y4hfdbxf2dbfvgs8k0qcx67jzjsay-python2.7-boto-2.47.0/lib/python2.7/site-packages','/nix/store/fxkrxrmbz73xxn7b975l23kz03s9gwzg-python2.7-urllib3-1.22/lib/python2.7/site-packages','/nix/store/6spad8r3v70sv231w1xjq26hn9gxzscy-python2.7-pyOpenSSL-17.5.0/lib/python2.7/site-packages','/nix/store/ws42f82li21s559ibyk1nkrkkskcf0k3-python2.7-certifi-2018.1.18/lib/python2.7/site-packages','/nix/store/cwkarmks7nghqb5av16j9nfpjmwc1ssq-python2.7-pysocks-1.6.6/lib/python2.7/site-packages','/nix/store/nn6kp8jknb0al4yhxa39vp9bhv4ssbvp-python2.7-chardet-3.0.4/lib/python2.7/site-packages','/nix/store/sywp8d82msj2cxcrwps1c6bzgxzmp3yq-python2.7-requests-2.18.4/lib/python2.7/site-packages','/nix/store/4l4kwjdjzybjfdx8673nrw9j2wjnpw7n-python2.7-httpretty-0.8.10/lib/python2.7/site-packages','/nix/store/lma2f5yk6aak4z6jsm7wvvm6idpx07fm-python2.7-netaddr-0.7.19/lib/python2.7/site-packages','/nix/store/wy2w3wbbzj4hawgxr4rb78qp6qp9macy-python2.7-dnspython-1.15.0/lib/python2.7/site-packages'], site._init_pathinfo());
```  

So the `site` import is what's failing for some reason, I suppose it's not evaluating the Python environment the same as it would be in my shell, which is able to execute flawlessly, so obviously can find the necessary modules

**Comment by radekg** [comment]
Is there any difference between:

```
$ /usr/local/bin/python2.7
>>> import site
```

and

```
$ /nix/store/573lgkld523h7hh5djfzfl8p66fcxawc-python-2.7.14/bin/python2.7
>>> import site
```

**Comment by cmacrae** [comment]
Nope, both behave as expected. I'll try and play around with some variables and see if I can alter the behaviour (thanks for your help so far! 👍)

**Comment by cmacrae** [comment]
Okay, so I've been doing a bit of debugging. As far as I can see, the wrapper script is being called correctly, and all the variables that need to be set are indeed being evaluated.  

That said, as you'll see in my expression for the host being provisioned, I have:  
```
extra_vars {
	ansible_python_interpreter = "/usr/local/bin/python2.7"
	ansible_bash_interpreter = "/usr/local/bin/bash"
      }
```
This is so Ansible knows where the **remote host** has Python and Bash. Am I using this correctly?  
I'm using these in lieu of `host_vars` or `group_vars` which is where I'd usually set this sort of thing.  

I just want to make sure I'm correct in my understanding and that this isn't what might be causing issues (commenting these out results in the Ansible execution unable to complete as it's not able to locate Python on the remote host, so I presume I'm correct).

**Comment by cmacrae** [comment]
Oh! Okay.... I got it working.  

I'm deploying an OpenBSD KVM instance on a private [Triton](https://www.joyent.com/triton/compute) cluster. In my Terraform expression for the instance, I was using `user_script = "pkg_add -z python-2.7" ` (this is run on boot) to install Python and ensure Ansible could run as a `provision` step.  

Before using your Ansible provisioner, I had a `remote-exec` provisioner which was just a loop waiting for Python to be installed:  
```
provisioner "remote-exec" {
    inline = [ 
      "while ! test -x /usr/local/bin/python2.7; do sleep 5 ; done"
      ]
    }
```

Of course, kinda superfluous to have these separated, so I just got rid of the `user_script` argument and moved the Python install into the `remote-exec` provisioner.  

After doing that... it all worked 🤔   

The only reasonable explanation I can think of is that I was just naively testing for the presence of `/usr/local/bin/python2.7`, which might have been present before the actual package install was complete. There might have been some post-install hooks that run to ensure the Python environment on OpenBSD is configured correctly.

Either way, I'm very happy!  


Thank you for your assistance in debugging, fortunately it looks like it's just down to some nuances in my setup 😁
